### ISSUE TITLE
a11y on Android: cut/paste not announced correctly

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#14331]
When text is selected in a text field and "cut" is chosen from the Local Context Menu, an announcement acknowledging the copy is expected (e.g. "cut \<text\>"). In Flutter that announcement is missing.

Similarly, when "paste" is chosen from the Local Context Menu, an announcement acknowledging the paste (e.g. "pasted  \<text\>") is expected. This one is also missing in Flutter.

**Comment by goderbauer** [comment]
Digging into the Flutter and TalkBack code: We do send the correct event `TYPE_VIEW_TEXT_CHANGED`, that would trigger the cut/paste announcement. However, we send it at the "wrong" time:

* When "cut" is selected from the Local Context Menu, the following code executes in TalkBack [1]:
   ```java
   EditTextActionHistory.getInstance().beforeCut();
   result = PerformActionUtils.performAction(mNode, AccessibilityNodeInfoCompat.ACTION_CUT);
   EditTextActionHistory.getInstance().afterCut();
   ```
* `beforeCut()` and `afterCut()` note down the current time.
* In the `performAction` method [2], Flutter's engine sends the request to do a cut over the async/sync boundary to the Framework in Dart and the method returns **before** the cut has actually been performed (that's necessary because processing happens asynchronously in the framework, yet we need to return synchronously from `performAction`).
* Once the cut has been performed on the Framework side in Dart, a semantics update is send back to the engine and we construct a `TYPE_VIEW_TEXT_CHANGED` event [3].
* That event is of course send with a timestamp older then the timestamp recorded by `afterCut()` above.
* Therefore, TalkBack does not identify this `TYPE_VIEW_TEXT_CHANGED` as belonging to a cut operation and does not produce the announcement expected from a cut operation.


[1] https://github.com/google/talkback/blob/12bdf2063e121a021f050c94cf5ebb2489c8af8a/src/main/java/com/android/talkback/menurules/RuleEditText.java#L256
[2] https://github.com/flutter/engine/blob/9a402241e9d679ff9dcd96712693b9df38cdc414/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L370
[3] https://github.com/flutter/engine/blob/9a402241e9d679ff9dcd96712693b9df38cdc414/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L529

**Comment by goderbauer** [comment]
Paste seems to follow a similar pattern.

**Comment by goderbauer** [comment]
TalkBack's behavior (expecting the `TYPE_VIEW_TEXT_CHANGED` event during the execution of `performAction`) is in line with what's documented for `performAction`: "Returns `True` if the action **was** performed." (emphasize mine).

We return true before the action was performed due to the async/sync gap and therefore violate the API's contract.

**Comment by goderbauer** [comment]
One hacky solution could be to send out a "fake" `TYPE_VIEW_TEXT_CHANGED` event with the anticipated change that would have the right timestamp and get us the correct announcements from TalkBack. We'd probably have to suppress the real event later on...

**Comment by goderbauer** [comment]
Side node: setting the timestamp of the `TYPE_VIEW_TEXT_CHANGED` event to a timestamp recorded during the execution of `performAction` using `setEventTime()` does not work either. The manually set time is overwritten by the actual time when the event is send.

**Comment by iapicca** [comment]
Hi @goderbauer is this issue still meant to be open?

**Comment by goderbauer** [comment]
It looks like this is still unresolved.

**Comment by darshankawar** [comment]
Tested this on latest Flutter dev channel and observed that although the  text selection menu when selected, is announced properly, (eg: `cut button, double tap to activate`, and similar for copy and paste), but when double tapped on any of these buttons, no announcement is made.
Currently, if `cut` is chosen, app announces it as `cleared textfield`. If `paste` is chosen, it announces as `a replaced b`.
In short, the issue still persists.

**Comment by huycozy** [comment]
This issue is reproducible on the latest `stable` and `master` channels. The behavior seems to be changed. For eg, when `paste` option is chosen, it will announce `pasted texts` and then announce all texts inside TextField.

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.3.9, on macOS 13.0 22A380 darwin-x64, locale en-VN)
    • Flutter version 3.3.9 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b8f7f1f986 (31 hours ago), 2022-11-23 06:43:51 +0900
    • Engine revision 8f2221fbef
    • Dart version 2.18.5
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (3 available)
    • SM T225 (mobile) • R9JT3004VRJ • android-arm64  • Android 12 (API 31)
    • macOS (desktop)  • macos       • darwin-x64     • macOS 13.0 22A380 darwin-x64
    • Chrome (web)     • chrome      • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

```bash
[!] Flutter (Channel master, 3.7.0-7.0.pre.18, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.7.0-7.0.pre.18 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f07db4018a (12 minutes ago), 2022-12-14 03:25:49 +0200
    • Engine revision 14194c40ec
    • Dart version 3.0.0 (build 3.0.0-19.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.54.0

[✓] Connected device (3 available)
    • SM T225 (mobile) • R9JT3004VRJ • android-arm64  • Android 12 (API 31)
    • macOS (desktop)  • macos       • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)     • chrome      • web-javascript • Google Chrome 108.0.5359.98

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```
</details>
